File: frontend/src/services/feed.ts
import { api } from './api';
import { FeedItem } from '../types';

export interface FeedTransaction {
  id: string;
  feed_id: string;
  feed_name: string;
  type: 'IN' | 'OUT';
  transaction_type?: 'STOCK_IN' | 'STOCK_OUT';
  amount: number;
  remaining_stock?: number;
  description: string;
  created_at: string;
  user_name: string;
}

export const feedService = {
  getFeeds: async () => {
    const res = await api.get<{ status: string; data: FeedItem[] }>('/feeds');
    return res.data;
  },

  getFeed: async (id: string) => {
    const res = await api.get<{ status: string; data: FeedItem }>(`/feeds/${id}`);
    return res.data;
  },

  saveFeed: async (feed: Partial<FeedItem>) => {
    const payload = {
      id: feed.id,
      nama: feed.nama,
      kategori: feed.kategori,
      stok: feed.stok,
      ambangBatas: feed.ambangBatas,
      nutrisi: feed.nutrisi || {
        protein: 0,
        karbohidrat: 0,
        lemak: 0,
        serat: 0,
        mineral: 0,
      },
      unit: feed.satuan || 'kg',
      price_per_kg: feed.hargaPerKg || 0,
    };
    return api.post<{ status: string; message: string; data: FeedItem }>('/feeds', payload);
  },

  deleteFeed: async (id: string) => {
    return api.delete<{ status: string; message: string }>(`/feeds/${id}`);
  },

  restockFeed: async (id: string, amount: number, description: string = '') => {
    return api.post<{ status: string; message: string }>(`/feeds/${id}/restock`, {
      amount,
      description,
    });
  },

  getTransactions: async (feedId?: string) => {
    const endpoint = feedId ? `/feeds/transactions?feed_id=${feedId}` : '/feeds/transactions';
    const res = await api.get<{ status: string; data: FeedTransaction[] }>(endpoint);
    return res.data;
  },
};

File: backend/app/models/feed.py
# app/models/feed.py
import uuid
from datetime import datetime
from app.utils.db import db

class Feed(db.Model):
    __tablename__ = 'feeds'

    id = db.Column(db.String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    name = db.Column(db.String(100), unique=True, nullable=False, index=True)
    category = db.Column(db.String(50), nullable=False)
    stock = db.Column(db.Float, default=0.0, nullable=False)
    min_threshold = db.Column(db.Float, default=5.0, nullable=False)
    protein = db.Column(db.Float, default=0.0, nullable=False)
    carbohydrate = db.Column(db.Float, default=0.0, nullable=False)
    fat = db.Column(db.Float, default=0.0, nullable=False)
    fiber = db.Column(db.Float, default=0.0, nullable=False)
    mineral = db.Column(db.Float, default=0.0, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    transactions = db.relationship('FeedTransaction', backref='feed', lazy=True, cascade="all, delete-orphan")

    def to_dict(self):
        return {
            'id': self.id,
            'nama': self.name,
            'kategori': self.category,
            'stok': self.stock,
            'ambangBatas': self.min_threshold,
            'nutrisi': {
                'protein': self.protein,
                'karbohidrat': self.carbohydrate,
                'lemak': self.fat,
                'serat': self.fiber,
                'mineral': self.mineral
            }
        }

class FeedTransaction(db.Model):
    __tablename__ = 'feed_transactions'

    id = db.Column(db.String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    feed_id = db.Column(db.String(50), db.ForeignKey('feeds.id', ondelete='CASCADE'), nullable=False)
    type = db.Column(db.Enum('IN', 'OUT', name='transaction_types'), nullable=False)
    amount = db.Column(db.Float, nullable=False)
    description = db.Column(db.Text, nullable=True)
    user_id = db.Column(db.String(50), db.ForeignKey('users.id', ondelete='SET NULL'), nullable=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

    def to_dict(self):
        return {
            'id': self.id,
            'feed_id': self.feed_id,
            'feed_name': self.feed.name if self.feed else None,
            'type': self.type,
            'amount': self.amount,
            'description': self.description,
            'user_id': self.user_id,
            'user_name': self.user.name if self.user else None,
            'created_at': self.created_at.isoformat()
        }
